File: setup/GithubActionsManager.py
import base64
import logging
from typing import Tuple

import requests
from requests.exceptions import RequestException
from nacl import encoding, public

class GitHubAPIError(Exception):
    """Custom exception raised for GitHub API interaction failures."""
    pass


class GitHubActionsManager:
    """Manages GitHub API interactions for workflows and repository secrets."""

    def __init__(self, logger: logging.Logger, token: str, repo: str | None, api_version: str = "2026-03-10") -> None:
        self.logger = logger
        self.repo = repo
        
        # Use a session to persist headers and optimize connection pooling
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": api_version
        })
        
        if not repo:
            self.user = self.fetch_user_info()
            self.repo = f"{self.user}/frc-train"
            
        self.base_url = f"https://api.github.com/repos/{self.repo}"
            
            
    def fetch_user_info(self):
        response = self.session.get("https://api.github.com/user")
        response.raise_for_status()
        return response.json().get('login')
            

    @staticmethod
    def encrypt_secret(public_key: str, secret_value: str) -> str:
        """Encrypt a string using the repository's public key via libsodium (pynacl)."""
        try:
            public_key_bytes = encoding.Base64Encoder.decode(public_key.encode("utf-8"))
            sealed_box = public.SealedBox(public.PublicKey(public_key_bytes))
            encrypted = sealed_box.encrypt(secret_value.encode("utf-8"))
            return base64.b64encode(encrypted).decode("utf-8")
        except Exception as e:
            raise GitHubAPIError(f"Failed to encrypt the secret via pynacl: {e}")

    def enable_repo_actions(self) -> None:
        """Enable GitHub Actions permissions for the repository."""
        url = f"{self.base_url}/actions/permissions"
        self.logger.info(f"Attempting to enable Actions permissions for '{self.repo}'...")
        
        try:
            response = self.session.put(url, json={"enabled": True})
            response.raise_for_status()  # Automatically raises RequestException for 4xx/5xx responses
            
            if response.status_code == 204:
                self.logger.info("Successfully enabled Actions permissions.")
            else:
                raise GitHubAPIError(f"Unexpected status code {response.status_code} while enabling actions.")
                
        except RequestException as e:
            raise GitHubAPIError(f"Network error or API rejection while enabling actions: {e}")

    def get_public_key(self) -> Tuple[str, str]:
        """Retrieve the repository's public key for secret encryption."""
        url = f"{self.base_url}/actions/secrets/public-key"
        self.logger.info(f"Retrieving public key for '{self.repo}'...")
        
        try:
            response = self.session.get(url)
            response.raise_for_status()
            
            data = response.json()
            return data["key_id"], data["key"]
            
        except RequestException as e:
            raise GitHubAPIError(f"Failed to retrieve public key: {e}")
        except KeyError as e:
            raise GitHubAPIError(f"Unexpected response format missing key: {e}")

    def set_secret(self, secret_name: str, secret_value: str) -> None:
        """Encrypt and set a repository secret."""
        # This will bubble up GitHubAPIError if it fails
        key_id, public_key = self.get_public_key()
        
        self.logger.info(f"Setting secret '{secret_name}' in '{self.repo}'...")
        encrypted_value = self.encrypt_secret(public_key, secret_value)
        
        url = f"{self.base_url}/actions/secrets/{secret_name}"
        payload = {
            "encrypted_value": encrypted_value,
            "key_id": key_id
        }
        
        try:
            response = self.session.put(url, json=payload)
            response.raise_for_status()
            
            if response.status_code in (201, 204):
                self.logger.info(f"Successfully configured the '{secret_name}' secret.")
            else:
                raise GitHubAPIError(f"Unexpected status code {response.status_code} while setting secret.")
                
        except RequestException as e:
            raise GitHubAPIError(f"Failed to set secret '{secret_name}': {e}")

    def dispatch_workflow(self, workflow_name: str, branch: str) -> None:
        """Dispatch a specific workflow on a target branch."""
        url = f"{self.base_url}/actions/workflows/{workflow_name}/dispatches"
        self.logger.info(f"Dispatching workflow '{workflow_name}' on branch '{branch}'...")
        
        try:
            response = self.session.post(url, json={"ref": branch})
            response.raise_for_status()
            
            if response.status_code == 200:
                self.logger.info("Successfully dispatched the workflow.")
            else:
                raise GitHubAPIError(f"Unexpected status code {response.status_code} while dispatching workflow.")
                
        except RequestException as e:
            raise GitHubAPIError(f"Failed to dispatch workflow: {e}")

File: setup/GithubActionsManager.js
const axios = require('axios');
const sodium = require('libsodium-wrappers');

class GitHubAPIError extends Error {
    constructor(message) {
        super(message);
        this.name = 'GitHubAPIError';
    }
}

class GitHubActionsManager {
    constructor(token, repo = null, apiVersion = "2026-03-10") {
        this.repo = repo;
        
        this.client = axios.create({
            baseURL: 'https://api.github.com',
            headers: {
                "Authorization": `Bearer ${token}`,
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": apiVersion
            }
        });
    }

    async init() {
        if (!this.repo) {
            const user = await this.fetchUserInfo();
            this.repo = `${user}/frc-train`;
        }
        this.baseUrl = `/repos/${this.repo}`;
        return this;
    }

    async fetchUserInfo() {
        try {
            const response = await this.client.get('/user');
            return response.data.login;
        } catch (error) {
            throw new GitHubAPIError(`Failed to fetch user info: ${error.message}`);
        }
    }

    static async encryptSecret(publicKey, secretValue) {
        try {
            await sodium.ready;
            const binkey = sodium.from_base64(publicKey, sodium.base64_variants.ORIGINAL);
            const binsec = sodium.from_string(secretValue);
            const encBytes = sodium.crypto_box_seal(binsec, binkey);
            return sodium.to_base64(encBytes, sodium.base64_variants.ORIGINAL);
        } catch (error) {
            throw new GitHubAPIError(`Failed to encrypt the secret via libsodium: ${error.message}`);
        }
    }

    async enableRepoActions() {
        const url = `${this.baseUrl}/actions/permissions`;
        console.log(`Attempting to enable Actions permissions for '${this.repo}'...`);
        
        try {
            const response = await this.client.put(url, { enabled: true });
            
            if (response.status === 204) {
                console.log("Successfully enabled Actions permissions.");
            } else {
                throw new GitHubAPIError(`Unexpected status code ${response.status} while enabling actions.`);
            }
        } catch (error) {
            throw new GitHubAPIError(`Network error or API rejection while enabling actions: ${error.message}`);
        }
    }

    async getPublicKey() {
        const url = `${this.baseUrl}/actions/secrets/public-key`;
        console.log(`Retrieving public key for '${this.repo}'...`);
        
        try {
            const response = await this.client.get(url);
            
            if (!response.data.key_id || !response.data.key) {
                throw new Error("Unexpected response format missing key.");
            }
            
            return {
                keyId: response.data.key_id,
                key: response.data.key
            };
        } catch (error) {
            throw new GitHubAPIError(`Failed to retrieve public key: ${error.message}`);
        }
    }

    async setSecret(secretName, secretValue) {
        const { keyId, key: publicKey } = await this.getPublicKey();
        
        console.log(`Setting secret '${secretName}' in '${this.repo}'...`);
        const encryptedValue = await GitHubActionsManager.encryptSecret(publicKey, secretValue);
        
        const url = `${this.baseUrl}/actions/secrets/${secretName}`;
        const payload = {
            encrypted_value: encryptedValue,
            key_id: keyId
        };
        
        try {
            const response = await this.client.put(url, payload);
            
            if (response.status === 201 || response.status === 204) {
                console.log(`Successfully configured the '${secretName}' secret.`);
            } else {
                throw new GitHubAPIError(`Unexpected status code ${response.status} while setting secret.`);
            }
        } catch (error) {
            throw new GitHubAPIError(`Failed to set secret '${secretName}': ${error.message}`);
        }
    }

    async dispatchWorkflow(workflowName, branch) {
        const url = `${this.baseUrl}/actions/workflows/${workflowName}/dispatches`;
        console.log(`Dispatching workflow '${workflowName}' on branch '${branch}'...`);
        
        try {
            const response = await this.client.post(url, { ref: branch });
            
            if (response.status === 200 || response.status === 204) {
                console.log("Successfully dispatched the workflow.");
            } else {
                throw new GitHubAPIError(`Unexpected status code ${response.status} while dispatching workflow.`);
            }
        } catch (error) {
            throw new GitHubAPIError(`Failed to dispatch workflow: ${error.message}`);
        }
    }
}

module.exports = {
    GitHubActionsManager,
    GitHubAPIError
};